internal/wasm/module.go

```diff
@@ -170,6 +173,274 @@ func (m *Module) TypeOfFunction(funcIdx Index) *FunctionType {
+			if index >= uint32(len(functions)) {
+				return fmt.Errorf("unknown function for export[%s]", name)
+			}
+		case ExternTypeGlobal:
+			if index >= uint32(len(globals)) {
+				return fmt.Errorf("unknown global for export[%s]", name)
+			}
+		case ExternTypeMemory:
+			if index != 0 || len(memories) == 0 {
+				return fmt.Errorf("unknown memory for export[%s]", name)
+			}
+		case ExternTypeTable:
+			if index >= uint32(len(tables)) {
+				return fmt.Errorf("unknown table for export[%s]", name)
+			}
+		}
+	}
+	return nil
+}
+
+func validateConstExpression(globals []*GlobalType, expr *ConstantExpression, expectedType ValueType) (err error) {
+	var actualType ValueType
+	r := bytes.NewBuffer(expr.Data)
+	switch expr.Opcode {
+	case OpcodeI32Const:
+		_, _, err = leb128.DecodeInt32(r)
+		if err != nil {
+			return fmt.Errorf("read i32: %w", err)
+		}
+		actualType = ValueTypeI32
+	case OpcodeI64Const:
+		_, _, err = leb128.DecodeInt64(r)
+		if err != nil {
+			return fmt.Errorf("read i64: %w", err)
+		}
+		actualType = ValueTypeI64
+	case OpcodeF32Const:
+		_, err = ieee754.DecodeFloat32(r)
+		if err != nil {
+			return fmt.Errorf("read f32: %w", err)
+		}
+		actualType = ValueTypeF32
+	case OpcodeF64Const:
+		_, err = ieee754.DecodeFloat64(r)
+		if err != nil {
+			return fmt.Errorf("read f64: %w", err)
+		}
+		actualType = ValueTypeF64
+	case OpcodeGlobalGet:
+		id, _, err := leb128.DecodeUint32(r)
+		if err != nil {
+			return fmt.Errorf("read index of global: %w", err)
+		}
+		if uint32(len(globals)) <= id {
+			return fmt.Errorf("global index out of range")
+		}
+		actualType = globals[id].ValType
+	default:
+		return fmt.Errorf("invalid opcode for const expression: 0x%x", expr.Opcode)
+	}
+
+	if actualType != expectedType {
+		return fmt.Errorf("const expression type mismatch")
+	}
+	return nil
+}
+
+func (m *Module) buildGlobalInstances(importedGlobals []*GlobalInstance) (globals []*GlobalInstance) {
+	for _, gs := range m.GlobalSection {
+		var gv uint64
+		// Global's initialization constant expression can only reference the imported globals.
+		// See the note on https://www.w3.org/TR/2019/REC-wasm-core-1-20191205/#constant-expressions%E2%91%A0
+		switch v := executeConstExpression(importedGlobals, gs.Init).(type) {
+		case int32:
+			gv = uint64(v)
+		case int64:
+			gv = uint64(v)
+		case float32:
+			gv = publicwasm.EncodeF32(v)
+		case float64:
+			gv = publicwasm.EncodeF64(v)
+		}
+		globals = append(globals, &GlobalInstance{
+			Type: gs.Type,
+			Val:  gv,
+		})
+	}
+	return
+}
+
+func (m *Module) buildFunctionInstances() (functions []*FunctionInstance) {
+	var functionNames NameMap
+	if m.NameSection != nil {
+		functionNames = m.NameSection.FunctionNames
+	}
+
+	importCount := m.ImportFuncCount()
+	n, nLen := 0, len(functionNames)
+	for codeIndex := range m.FunctionSection {
+		funcIdx := Index(importCount + uint32(len(functions)))
+		// Seek to see if there's a better name than "unknown"
+		name := "unknown"
```

Comment:
probably better to default to $funcIdx and set that after the loop so we don't allocate unless we found no name.